05.ls/ls.rb

```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require 'optparse'
+require 'etc'
+
+def main
+  options = ARGV.getopts('a', 'l', 'r')
+  file_list = if options['a']
+                Dir.glob('*', File::FNM_DOTMATCH, base: ARGV[0])
+              else
+                Dir.glob('*', base: ARGV[0])
+              end
+
+  file_list.reverse! if options['r']
+  if options['l']
+    listing_option_l(file_list)
+  else
+    listing(file_list)
+  end
+end
+
+def listing(file_list)
+  # -l オプションなしのファイルリスト表示
+  col = 3 # 3列表示
+  file_count = file_list.size
+  row = (file_count % col).positive? ? file_count / col + 1 : file_count / col
+  max_values = values_max_length(file_list)
+
+  listing_text = ''
+  row.times do |i|
+    col.times do |ii|
+      listing_text += mb_ljust(file_list[i + ii * row], max_values[0] + 1) if i + ii * row < file_count
+    end
+    listing_text += "\n"
+  end
+  puts listing_text
+end
+
+def listing_option_l(file_list)
+  # -l オプション指定時のファイルリスト表示
+  # -rwxr--r--  1 ogaworks  staff   409B  6 22 22:59 ls.rb
+  xattr_file_list = extended_attributes(ARGV)
+  max_values = values_max_length(file_list)
+  file_list.each do |file|
+    file_path = "#{ARGV[0]}#{file}"
+    stat = File.lstat(file_path)
+    list_text = "#{file_type(stat.mode)}#{file_permission(stat.mode)}"
+    list_text += xattr_file_list&.include?(file) && file_type(stat.mode) == '-' ? '@' : ' '
+    list_text += "#{stat.nlink.to_s.rjust(3)} "
+    list_text += "#{Etc.getpwuid(File.lstat(file_path).uid).name.ljust(max_values[2])} #{Etc.getgrgid(File.lstat(file_path).gid).name.rjust(max_values[3] + 1)}"
+    list_text += "#{stat.size.to_s.rjust(max_values[1] + 2)} "
+    list_text += "#{stat.mtime.strftime('%-m %e %H:%M').rjust(11)} "
+    file_name = file
+    list_text += if file_type(stat.mode) == 'l'
+                   "#{file_name} -> #{File.readlink(file_path)}"
+                 else
+                   file_name
+                 end
+    puts list_text
+  end
+end
+
+def mb_ljust(file_name, width)
+  char_count = file_name.each_char.map { |c| c.bytesize == 1 ? 1 : 2 }.inject(0, :+)
```

Comment:
```suggestion
  char_count = file_name.each_char.sum { |c| c.bytesize == 1 ? 1 : 2 }
```
でよさそうです。
https://docs.ruby-lang.org/ja/latest/method/Enumerable/i/sum.html